From 3a02f977b587dd1f00aee6a38f557ac86bb7d846 Mon Sep 17 00:00:00 2001
From: John Reck <jreck@google.com>
Date: Tue, 10 Mar 2026 14:29:24 -0400
Subject: [PATCH] Update to DNG SDK 1.7.1 2502

Test: make
Flag: EXEMPT BUGFIX
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:51fac39e7e2a05b2c5fa54108035d0140a5bc86c
Merged-In: I98050384600e0f4ac77f4378e00eee9be4830aa7
Change-Id: I98050384600e0f4ac77f4378e00eee9be4830aa7
Bug: 491543354
---
 README.version                |   4 +-
 source/dng_big_table.cpp      |   2 +-
 source/dng_camera_profile.cpp |   8 ++
 source/dng_flags.h            |   4 +-
 source/dng_globals.h          |   2 +-
 source/dng_negative.cpp       |   2 +-
 source/dng_negative.h         |   2 +-
 source/dng_pixel_buffer.cpp   | 182 +++++++++++++++++++++++-----------
 source/dng_point.h            |  23 +++++
 source/dng_render.cpp         | 150 +++++++++++++++-------------
 source/dng_utils.cpp          |   4 +
 11 files changed, 244 insertions(+), 139 deletions(-)

diff --git a/README.version b/README.version
index 6cb2483..efc65ae 100644
--- a/README.version
+++ b/README.version
@@ -1,5 +1,5 @@
-URL: http://download.adobe.com/pub/adobe/dng/dng_sdk_1_4.zip
-Version: 1.4.0
+URL: https://download.adobe.com/pub/adobe/dng/dng_sdk_1_7_1_2502_20260303.zip
+Version: 1.7.1
 License: BSD like License
 License File: LICENSE
 BugComponent: 174430
diff --git a/source/dng_big_table.cpp b/source/dng_big_table.cpp
index e8988a9..e1716f8 100644
--- a/source/dng_big_table.cpp
+++ b/source/dng_big_table.cpp
@@ -3122,7 +3122,7 @@ void dng_image_table::PutCompressedStream (dng_stream &stream,
 		
 		tempImage.Reset (tiffImage->Clone ());
 		
-		tempImage->Offset (dng_point (0, 0) - fImage->Bounds ().TL ());
+		tempImage->Offset (-fImage->Bounds ().TL ());
 		
 		tiffImage = tempImage.Get ();
 		
diff --git a/source/dng_camera_profile.cpp b/source/dng_camera_profile.cpp
index d5dc962..6c716b2 100644
--- a/source/dng_camera_profile.cpp
+++ b/source/dng_camera_profile.cpp
@@ -1850,6 +1850,8 @@ void dng_camera_profile::SetFourColorBayer ()
 	fForwardMatrix1.Clear ();
 	fForwardMatrix2.Clear ();
 	fForwardMatrix3.Clear ();
+
+	ClearFingerprint ();
 	
 	}
 
@@ -2039,6 +2041,8 @@ void dng_camera_profile::SetProfileGainTableMap
 	{
 	
 	fProfileGainTableMap = gainTableMap;
+
+	ClearFingerprint ();
 	
 	}
 
@@ -2112,6 +2116,8 @@ void dng_camera_profile::SetMaskedRGBTables
 	{
 	
 	fMaskedRGBTables = maskedRGBTables;
+
+	ClearFingerprint ();
 	
 	}
 
@@ -2122,6 +2128,8 @@ void dng_camera_profile::SetMaskedRGBTables
 	{
 	
 	fMaskedRGBTables.reset (maskedRGBTables.Release ());
+
+	ClearFingerprint ();
 	
 	}
 
diff --git a/source/dng_flags.h b/source/dng_flags.h
index a632829..4fde43e 100644
--- a/source/dng_flags.h
+++ b/source/dng_flags.h
@@ -27,8 +27,8 @@
 
 #define kDNGSDK_VersionString "1.7.1"
 
-#define kDNGSDK_BuildVersion 2471
-#define kDNGSDK_BuildString  "2471"
+#define kDNGSDK_BuildVersion 2502
+#define kDNGSDK_BuildString  "2502"
 
 #define kDNGSDK_GetInfoVersion	"1.7.1 (" kDNGSDK_BuildString ")"
 
diff --git a/source/dng_globals.h b/source/dng_globals.h
index d2bacfd..e5b7be0 100644
--- a/source/dng_globals.h
+++ b/source/dng_globals.h
@@ -52,7 +52,7 @@ extern bool gDNGUseFakeTimeZonesInXMP;
 /*****************************************************************************/
 
 // Stream block size.  Choose a size that the OS likes for file system
-// efficent read/write alignment.
+// efficient read/write alignment.
 
 extern uint32 gDNGStreamBlockSize;
 
diff --git a/source/dng_negative.cpp b/source/dng_negative.cpp
index 2126416..8612003 100644
--- a/source/dng_negative.cpp
+++ b/source/dng_negative.cpp
@@ -6844,7 +6844,7 @@ void dng_negative::ConvertToProxy (dng_host &host,
 			
 		// Figure out the requested proxy pixel size.
 		
-		real64 aspectRatio = AspectRatio ();
+		real64 aspectRatio = BaseAspectRatio ();
 		
 		dng_point newSize (proxySize, proxySize);
 		
diff --git a/source/dng_negative.h b/source/dng_negative.h
index f470f00..9c8524a 100644
--- a/source/dng_negative.h
+++ b/source/dng_negative.h
@@ -1544,7 +1544,7 @@ class dng_negative
 		
 		/// Default cropped image aspect ratio.
 		
-		real64 AspectRatio () const
+		real64 BaseAspectRatio () const
 			{
 			return SquareWidth	() /
 				   SquareHeight ();
diff --git a/source/dng_pixel_buffer.cpp b/source/dng_pixel_buffer.cpp
index 990799b..1c65ee0 100644
--- a/source/dng_pixel_buffer.cpp
+++ b/source/dng_pixel_buffer.cpp
@@ -15,6 +15,8 @@
 #include "dng_tag_values.h"
 #include "dng_utils.h"
 
+#include <climits>
+
 /*****************************************************************************/
 
 static bool SafeUint32ToInt32Mult (uint32 arg1, 
@@ -31,6 +33,55 @@ static bool SafeUint32ToInt32Mult (uint32 arg1,
 			ConvertUint32ToInt32 (uint32_result, 
 								  result));
 
+		}
+
+/*****************************************************************************/
+
+static int64 ComputeRebaseDelta (uint32 count,
+								 int32 step,
+								 uint32 pixelSize)
+	{
+
+	const int64 countMinusOne = static_cast<int64> (count) - 1;
+
+	return SafeInt64Mult (countMinusOne,
+						  static_cast<int64> (step),
+						  static_cast<int64> (pixelSize));
+
+	}
+
+/*****************************************************************************/
+
+static const void * OffsetPointer (const void *ptr,
+								   int64 delta)
+	{
+
+	return static_cast<const void *> (static_cast<const uint8 *> (ptr) + delta);
+
+	}
+
+/*****************************************************************************/
+
+static void * OffsetPointer (void *ptr,
+							 int64 delta)
+	{
+
+	return static_cast<void *> (static_cast<uint8 *> (ptr) + delta);
+
+	}
+
+/*****************************************************************************/
+
+static int32 NegateStepChecked (int32 step)
+	{
+
+	if (step == INT32_MIN)
+		{
+		ThrowOverflow ("OptimizeOrder step negate");
+		}
+
+	return -step;
+
 	}
 
 /*****************************************************************************/
@@ -49,6 +100,11 @@ void OptimizeOrder (const void *&sPtr,
 					int32 &dStep1,
 					int32 &dStep2)
 	{
+
+	if (count0 == 0 || count1 == 0 || count2 == 0)
+		{
+		return;
+		}
 	
 	uint32 step0;
 	uint32 step1;
@@ -56,56 +112,62 @@ void OptimizeOrder (const void *&sPtr,
 	
 	// Optimize the order for the data that is most spread out.
 							   
-	uint32 sRange = Abs_int32 (sStep0) * (count0 - 1) +
-					Abs_int32 (sStep1) * (count1 - 1) +
-					Abs_int32 (sStep2) * (count2 - 1);
+	uint64 sRange = static_cast<uint64> (Abs_int32 (sStep0)) *
+					static_cast<uint64> (count0 - 1) +
+					static_cast<uint64> (Abs_int32 (sStep1)) *
+					static_cast<uint64> (count1 - 1) +
+					static_cast<uint64> (Abs_int32 (sStep2)) *
+					static_cast<uint64> (count2 - 1);
 							   
-	uint32 dRange = Abs_int32 (dStep0) * (count0 - 1) +
-					Abs_int32 (dStep1) * (count1 - 1) +
-					Abs_int32 (dStep2) * (count2 - 1);
+	uint64 dRange = static_cast<uint64> (Abs_int32 (dStep0)) *
+					static_cast<uint64> (count0 - 1) +
+					static_cast<uint64> (Abs_int32 (dStep1)) *
+					static_cast<uint64> (count1 - 1) +
+					static_cast<uint64> (Abs_int32 (dStep2)) *
+					static_cast<uint64> (count2 - 1);
 							   
 	if (dRange >= sRange)
 		{						
 	
 		if (dStep0 < 0)
 			{
-			
-			sPtr = (const void *)
-				   (((const uint8 *) sPtr) + (int32)(count0 - 1) * sStep0 * (int32)sPixelSize);
-				   
-			dPtr = (void *)
-				   (((uint8 *) dPtr) + (int32)(count0 - 1) * dStep0 * (int32)dPixelSize);
-				   
-			sStep0 = -sStep0;
-			dStep0 = -dStep0;
+
+			const int64 sDelta0 = ComputeRebaseDelta (count0, sStep0, sPixelSize);
+			const int64 dDelta0 = ComputeRebaseDelta (count0, dStep0, dPixelSize);
+
+			sPtr = OffsetPointer (sPtr, sDelta0);
+			dPtr = OffsetPointer (dPtr, dDelta0);
+
+			sStep0 = NegateStepChecked (sStep0);
+			dStep0 = NegateStepChecked (dStep0);
 			
 			}
 		
 		if (dStep1 < 0)
 			{
-			
-			sPtr = (const void *)
-				   (((const uint8 *) sPtr) + (int32)(count1 - 1) * sStep1 * (int32)sPixelSize);
-				   
-			dPtr = (void *)
-				   (((uint8 *) dPtr) + (int32)(count1 - 1) * dStep1 * (int32)dPixelSize);
-				   
-			sStep1 = -sStep1;
-			dStep1 = -dStep1;
+
+			const int64 sDelta1 = ComputeRebaseDelta (count1, sStep1, sPixelSize);
+			const int64 dDelta1 = ComputeRebaseDelta (count1, dStep1, dPixelSize);
+
+			sPtr = OffsetPointer (sPtr, sDelta1);
+			dPtr = OffsetPointer (dPtr, dDelta1);
+
+			sStep1 = NegateStepChecked (sStep1);
+			dStep1 = NegateStepChecked (dStep1);
 			
 			}
 		
 		if (dStep2 < 0)
 			{
-			
-			sPtr = (const void *)
-				   (((const uint8 *) sPtr) + (int32)(count2 - 1) * sStep2 * (int32)sPixelSize);
-				   
-			dPtr = (void *)
-				   (((uint8 *) dPtr) + (int32)(count2 - 1) * dStep2 * (int32)dPixelSize);
-				   
-			sStep2 = -sStep2;
-			dStep2 = -dStep2;
+
+			const int64 sDelta2 = ComputeRebaseDelta (count2, sStep2, sPixelSize);
+			const int64 dDelta2 = ComputeRebaseDelta (count2, dStep2, dPixelSize);
+
+			sPtr = OffsetPointer (sPtr, sDelta2);
+			dPtr = OffsetPointer (dPtr, dDelta2);
+
+			sStep2 = NegateStepChecked (sStep2);
+			dStep2 = NegateStepChecked (dStep2);
 			
 			}
 		
@@ -120,43 +182,43 @@ void OptimizeOrder (const void *&sPtr,
 	
 		if (sStep0 < 0)
 			{
-			
-			sPtr = (const void *)
-				   (((const uint8 *) sPtr) + (int32)(count0 - 1) * sStep0 * (int32)sPixelSize);
-				   
-			dPtr = (void *)
-				   (((uint8 *) dPtr) + (int32)(count0 - 1) * dStep0 * (int32)dPixelSize);
-				   
-			sStep0 = -sStep0;
-			dStep0 = -dStep0;
+
+			const int64 sDelta0 = ComputeRebaseDelta (count0, sStep0, sPixelSize);
+			const int64 dDelta0 = ComputeRebaseDelta (count0, dStep0, dPixelSize);
+
+			sPtr = OffsetPointer (sPtr, sDelta0);
+			dPtr = OffsetPointer (dPtr, dDelta0);
+
+			sStep0 = NegateStepChecked (sStep0);
+			dStep0 = NegateStepChecked (dStep0);
 			
 			}
 		
 		if (sStep1 < 0)
 			{
-			
-			sPtr = (const void *)
-				   (((const uint8 *) sPtr) + (int32)(count1 - 1) * sStep1 * (int32)sPixelSize);
-				   
-			dPtr = (void *)
-				   (((uint8 *) dPtr) + (int32)(count1 - 1) * dStep1 * (int32)dPixelSize);
-				   
-			sStep1 = -sStep1;
-			dStep1 = -dStep1;
+
+			const int64 sDelta1 = ComputeRebaseDelta (count1, sStep1, sPixelSize);
+			const int64 dDelta1 = ComputeRebaseDelta (count1, dStep1, dPixelSize);
+
+			sPtr = OffsetPointer (sPtr, sDelta1);
+			dPtr = OffsetPointer (dPtr, dDelta1);
+
+			sStep1 = NegateStepChecked (sStep1);
+			dStep1 = NegateStepChecked (dStep1);
 			
 			}
 		
 		if (sStep2 < 0)
 			{
-			
-			sPtr = (const void *)
-				   (((const uint8 *) sPtr) + (int32)(count2 - 1) * sStep2 * (int32)sPixelSize);
-				   
-			dPtr = (void *)
-				   (((uint8 *) dPtr) + (int32)(count2 - 1) * dStep2 * (int32)dPixelSize);
-				   
-			sStep2 = -sStep2;
-			dStep2 = -dStep2;
+
+			const int64 sDelta2 = ComputeRebaseDelta (count2, sStep2, sPixelSize);
+			const int64 dDelta2 = ComputeRebaseDelta (count2, dStep2, dPixelSize);
+
+			sPtr = OffsetPointer (sPtr, sDelta2);
+			dPtr = OffsetPointer (dPtr, dDelta2);
+
+			sStep2 = NegateStepChecked (sStep2);
+			dStep2 = NegateStepChecked (dStep2);
 			
 			}
 		
diff --git a/source/dng_point.h b/source/dng_point.h
index 20377a3..3b37b27 100644
--- a/source/dng_point.h
+++ b/source/dng_point.h
@@ -178,6 +178,29 @@ inline dng_point_real64 operator- (const dng_point_real64 &a,
 
 /*****************************************************************************/
 
+inline dng_point operator- (const dng_point &p)
+				  
+				  
+	{
+	
+	return dng_point (SafeInt32Sub (0, p.v),
+					  SafeInt32Sub (0, p.h));
+	
+	}
+
+/*****************************************************************************/
+
+inline dng_point_real64 operator- (const dng_point_real64 &p)
+				  
+				  
+	{
+	
+	return dng_point_real64 (-p.v, -p.h);
+	
+	}
+
+/*****************************************************************************/
+
 inline real64 Distance (const dng_point_real64 &a,
 						const dng_point_real64 &b)
 				  
diff --git a/source/dng_render.cpp b/source/dng_render.cpp
index db54a9c..bb11e17 100644
--- a/source/dng_render.cpp
+++ b/source/dng_render.cpp
@@ -1505,16 +1505,24 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 								   dng_pixel_buffer &srcBuffer,
 								   dng_pixel_buffer &dstBuffer)
 	{
-	
+
+	// This method expects and requires srcBuffer and dstBuffer areas
+	// to be the same SIZE. The work for each row will be done in buffers
+	// mapping srcArea and the resulting data for each row will be
+	// translated and copied to dstArea in dstBuffer at the final step.
+
 	dng_rect srcArea = srcBuffer.fArea;
 	dng_rect dstArea = dstBuffer.fArea;
-	
-	uint32 srcCols = srcArea.W ();
+
+	DNG_REQUIRE (srcArea.Size () == dstArea.Size (), "area size mismatch");
+
+	const uint32 cols = srcArea.W ();
+	const int32 colsSigned = ConvertUint32ToInt32 (cols);
 	
 	real32 *tPtrR = fTempBuffer [threadIndex]->Buffer_real32 ();
 	
-	real32 *tPtrG = tPtrR + srcCols;
-	real32 *tPtrB = tPtrG + srcCols;
+	real32 *tPtrG = tPtrR + cols;
+	real32 *tPtrB = tPtrG + cols;
 	
 	dng_pixel_buffer maskBuffer;
 		
@@ -1552,7 +1560,7 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 		buffer.fPlanes	  = fNumTableTransformPlanes;
 
 		buffer.fColStep	  = 1;
-		buffer.fPlaneStep = int32 (buffer.fArea.W ());
+		buffer.fPlaneStep = ConvertUint32ToInt32 (buffer.fArea.W ());
 		buffer.fRowStep	  = SafeInt32Mult (buffer.fPlaneStep,
 										   int32 (buffer.fPlanes));
 
@@ -1595,8 +1603,8 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 
 				dng_pixel_buffer tempMaskBuffer = buffer;
 
-				tempMaskBuffer.fData = tempMaskBuffer.DirtyPixel (dstArea.t,
-																  dstArea.l,
+				tempMaskBuffer.fData = tempMaskBuffer.DirtyPixel (srcArea.t,
+																  srcArea.l,
 																  dstMaskPlane);
 
 				tempMaskBuffer.fPlanes = 1;
@@ -1624,26 +1632,24 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 			if (backgroundIndex >= 0)
 				{
 
-				const int32 cols = (int32) dstArea.W ();
-
 				const int32 mPlaneStep = buffer.PlaneStep ();
 
-				for (int32 row = dstArea.t; row < dstArea.b; row++)
+				for (int32 row = srcArea.t; row < srcArea.b; row++)
 					{
 
 					// Initialize mask pointer to first plane for this row.
 
 					const real32 *mPtr = buffer.ConstPixel_real32 (row,
-																   dstArea.l,
+																   srcArea.l,
 																   0);
 
 					// We will write the result into the background mask plane.
 
 					real32 *dPtr = buffer.DirtyPixel_real32 (row,
-															 dstArea.l,
+															 srcArea.l,
 															 backgroundIndex);
 
-					for (int32 j = 0; j < cols; j++)
+					for (int32 j = 0; j < colsSigned; j++)
 						{
 
 						real32 mSum = 0.0f;
@@ -1711,8 +1717,8 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 
 				dng_pixel_buffer tempMaskBuffer = tempRGBTablesBuffer;
 
-				tempMaskBuffer.fData = tempMaskBuffer.DirtyPixel (dstArea.t,
-																  dstArea.l,
+				tempMaskBuffer.fData = tempMaskBuffer.DirtyPixel (srcArea.t,
+																  srcArea.l,
 																  dstMaskPlane);
 
 				tempMaskBuffer.fPlanes = 1;
@@ -1726,6 +1732,39 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 		
 		} // RGBTables
 
+	// Select which ProfileGainTableMap to use in row loop below.
+
+	std::shared_ptr<const dng_gain_table_map> pgtm;
+
+	// If the profile has a ProfileGainTableMap, then just use that.
+
+	const dng_camera_profile_id &profileID = fParams.CameraProfileID ();
+
+	dng_camera_profile profile;
+
+	if (fNegative.GetProfileByID (profileID, profile) &&
+		profile.HasProfileGainTableMap ())
+		{
+
+		pgtm = profile.ShareProfileGainTableMap ();
+
+		}
+
+	// Otherwise fall back to the PGTM attached to the negative itself,
+	// which corresponds to either ProfileGainTableMap2 in IFD 0 or
+	// ProfileGainTableMap in Raw IFD.
+
+	else if (fNegative.HasProfileGainTableMap ())
+		{
+
+		pgtm = fNegative.ShareProfileGainTableMap ();
+
+		}
+
+	// Compute exposureWeightGain for use inside row loop below.
+
+	const real32 exposureWeightGain = (real32) pow (2.0, fBaselineExposure);
+
 	// Process each row of the image.
 	
 	for (int32 srcRow = srcArea.t; srcRow < srcArea.b; srcRow++)
@@ -1744,7 +1783,7 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 
 				DoBaseline1DTable (sPtr,
 								   sPtr,
-								   srcCols,
+								   cols,
 								   fZeroOffsetRamp);
 					
 				}
@@ -1767,9 +1806,9 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 				// For monochrome cameras, this just requires copying
 				// the data into all three color channels.
 				
-				DoCopyBytes (sPtrA, tPtrR, srcCols * (uint32) sizeof (real32));
-				DoCopyBytes (sPtrA, tPtrG, srcCols * (uint32) sizeof (real32));
-				DoCopyBytes (sPtrA, tPtrB, srcCols * (uint32) sizeof (real32));
+				DoCopyBytes (sPtrA, tPtrR, cols * (uint32) sizeof (real32));
+				DoCopyBytes (sPtrA, tPtrG, cols * (uint32) sizeof (real32));
+				DoCopyBytes (sPtrA, tPtrB, cols * (uint32) sizeof (real32));
 				
 				}
 				
@@ -1788,7 +1827,7 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 										tPtrR,
 										tPtrG,
 										tPtrB,
-										srcCols,
+										cols,
 										fCameraWhite,
 										fCameraToRGB);
 					
@@ -1811,7 +1850,7 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 										 tPtrR,
 										 tPtrG,
 										 tPtrB,
-										 srcCols,
+										 cols,
 										 fCameraWhite,
 										 fCameraToRGB);
 					
@@ -1828,7 +1867,7 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 										 tPtrR,
 										 tPtrG,
 										 tPtrB,
-										 srcCols,
+										 cols,
 										 *fHueSatMap.Get (),
 										 fHueSatMapEncode.Get (),
 										 fHueSatMapDecode.Get (),
@@ -1850,49 +1889,18 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 		// the correct result by effectively scaling the MapInputWeights
 		// parameter by the baseline exposure.
 
-		// Select which ProfileGainTableMap to use.
-
-		std::shared_ptr<const dng_gain_table_map> pgtm;
-
-		// If the profile has a ProfileGainTableMap, then just use that.
-
-		const dng_camera_profile_id &profileID = fParams.CameraProfileID ();
-		
-		dng_camera_profile profile;
-
-		if (fNegative.GetProfileByID (profileID, profile) &&
-			profile.HasProfileGainTableMap ())
-			{
-			
-			pgtm = profile.ShareProfileGainTableMap ();
-			
-			}
-
-		// Otherwise fall back to the PGTM attached to the negative itself,
-		// which corresponds to either ProfileGainTableMap2 in IFD 0 or
-		// ProfileGainTableMap in Raw IFD.
-
-		else if (fNegative.HasProfileGainTableMap ())
-			{
-			
-			pgtm = fNegative.ShareProfileGainTableMap ();
-			
-			}
-
 		if (pgtm)
 			{
 
 			const dng_rect activeArea = fNegative.Stage3Image ()->Bounds ();
 
-			const real32 exposureWeightGain = (real32) pow (2.0, fBaselineExposure);
-
 			DoBaselineProfileGainTableMap (tPtrR, // src
 										   tPtrG,
 										   tPtrB,
 										   tPtrR, // dst
 										   tPtrG,
 										   tPtrB,
-										   srcCols,	  // columns
+										   cols,	  // columns
 										   srcRow,	  // top of tile
 										   srcArea.l, // left of tile
 										   activeArea,
@@ -1909,19 +1917,19 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 
 			DoBaseline1DFunction (tPtrR,
 								  tPtrR,
-								  srcCols,
+								  cols,
 								  *fExposureRamp,
 								  fSupportOverrange);
 								
 			DoBaseline1DFunction (tPtrG,
 								  tPtrG,
-								  srcCols,
+								  cols,
 								  *fExposureRamp,
 								  fSupportOverrange);
 
 			DoBaseline1DFunction (tPtrB,
 								  tPtrB,
-								  srcCols,
+								  cols,
 								  *fExposureRamp,
 								  fSupportOverrange);
 
@@ -1938,7 +1946,7 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 								 tPtrR,
 								 tPtrG,
 								 tPtrB,
-								 srcCols,
+								 cols,
 								 *fLookTable.Get (),
 								 fLookTableEncode.Get (),
 								 fLookTableDecode.Get (),
@@ -1955,7 +1963,7 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 										tPtrR,
 										tPtrG,
 										tPtrB,
-										srcCols,
+										cols,
 										fToneCurve,
 										*fToneCurveSlopeExtension);
 
@@ -1966,7 +1974,7 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 							   tPtrR,
 							   tPtrG,
 							   tPtrB,
-							   srcCols,
+							   cols,
 							   fToneCurve);
 
 		// Apply RGBTables, if any.
@@ -1974,14 +1982,14 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 		if (fRGBTablesData)
 			{
 
-			dng_rect dstRowArea (srcRow,
+			dng_rect srcRowArea (srcRow,
 								 srcArea.l,
 								 srcRow + 1,
 								 srcArea.r);
 
 			dng_pixel_buffer buffer;
 
-			buffer.fArea	  = dstRowArea;
+			buffer.fArea	  = srcRowArea;
 
 			buffer.fPlanes	  = 3;
 
@@ -2015,14 +2023,14 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 								 tPtrG,
 								 tPtrB,
 								 dPtrG,
-								 srcCols,
+								 cols,
 								 fRGBtoFinal,
 								 fSupportOverrange);
 
 			if (fEncodeGamma.Get ())
 				DoBaseline1DTable (dPtrG,
 								   dPtrG,
-								   srcCols,
+								   cols,
 								   *fEncodeGamma);
 								
 			}
@@ -2043,7 +2051,7 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 								dPtrR,
 								dPtrG,
 								dPtrB,
-								srcCols,
+								cols,
 								fRGBtoFinal,
 								fSupportOverrange);
 
@@ -2052,17 +2060,17 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 			
 				DoBaseline1DTable (dPtrR,
 								   dPtrR,
-								   srcCols,
+								   cols,
 								   *fEncodeGamma);
 
 				DoBaseline1DTable (dPtrG,
 								   dPtrG,
-								   srcCols,
+								   cols,
 								   *fEncodeGamma);
 
 				DoBaseline1DTable (dPtrB,
 								   dPtrB,
-								   srcCols,
+								   cols,
 								   *fEncodeGamma);
 
 				}
@@ -2083,7 +2091,7 @@ void dng_render_task::ProcessArea (uint32 threadIndex,
 															dstArea.l,
 															dstPlane);
 					
-				for (uint32 col = 0; col < srcCols; col++)
+				for (uint32 col = 0; col < cols; col++)
 					{
 					
 					// White Matte
@@ -2189,7 +2197,7 @@ dng_image * dng_render::Render ()
 		if (Max_uint32 (dstSize.h, dstSize.v) > MaximumSize ())
 			{
 			
-			real64 ratio = fNegative.AspectRatio ();
+			real64 ratio = fNegative.BaseAspectRatio ();
 			
 			if (ratio >= 1.0)
 				{
diff --git a/source/dng_utils.cpp b/source/dng_utils.cpp
index 079a08f..df1d76e 100644
--- a/source/dng_utils.cpp
+++ b/source/dng_utils.cpp
@@ -208,6 +208,10 @@ void dng_show_message (const char *s)
 
 /*****************************************************************************/
 
+#if (qMacOS || qiPhone)
+__attribute__((__format__ (__printf__, 1, 2)))
+#endif
+
 void dng_show_message_f (const char *fmt, ... )
 	{
 	
-- 
2.55.0.897.gb25b4bd76c-goog

